File: quest/Assets/Orion/Runtime/Flight/RoutePath.cs
using System.Collections.Generic;
using UnityEngine;

namespace Orion.Flight
{
    /// <summary>A polyline in world space with arc-length lookup. One path per leg, joined into a
    /// route, so "how far along" is a single number of metres.</summary>
    public class RoutePath
    {
        public readonly List<Vector3> Pts;
        public readonly float Length;
        readonly float[] cum;

        public RoutePath(List<Vector3> pts)
        {
            Pts = pts;
            cum = new float[Mathf.Max(1, pts.Count)];
            for (int i = 1; i < pts.Count; i++) cum[i] = cum[i - 1] + Vector3.Distance(pts[i], pts[i - 1]);
            Length = cum[cum.Length - 1];
        }

        /// <summary>Where `s` metres falls, counted in points: 2.5 is halfway between the third and the fourth.</summary>
        public float IndexAt(float s)
        {
            int n = Pts.Count;
            if (n < 2 || s <= 0) return 0;
            if (s >= Length) return n - 1;
            int lo = 0, hi = n - 1;
            while (hi - lo > 1) { int mid = (lo + hi) >> 1; if (cum[mid] <= s) lo = mid; else hi = mid; }
            float seg = cum[hi] - cum[lo];
            return lo + (seg > 0 ? (s - cum[lo]) / seg : 0);
        }

        public Vector3 At(float s)
        {
            if (Pts.Count == 0) return Vector3.zero;
            float f = IndexAt(s);
            int i = (int)f;
            return i + 1 < Pts.Count ? Vector3.LerpUnclamped(Pts[i], Pts[i + 1], f - i) : Pts[i];
        }

        /// <summary>Horizontal unit direction of travel around s, or `fallback` if the path is degenerate there.</summary>
        public Vector3 Heading(float s, float back, float ahead, Vector3 fallback)
        {
            Vector3 a = At(s - back), b = At(s + ahead);
            var d = new Vector3(b.x - a.x, 0, b.z - a.z);
            return d.sqrMagnitude < 1e-6f ? fallback : d.normalized;
        }
    }
}

File: src/fly/routePath.ts
import { Vector3 } from 'three'

/** A polyline in world space with arc-length lookup. One path per leg, joined
    into a route, so "how far along" is a single number of metres. */
export class Path {
  readonly pts: Vector3[]
  readonly cum: number[]
  readonly length: number

  constructor(pts: Vector3[]) {
    this.pts = pts
    this.cum = [0]
    for (let i = 1; i < pts.length; i++) this.cum.push(this.cum[i - 1] + pts[i].distanceTo(pts[i - 1]))
    this.length = this.cum[this.cum.length - 1] ?? 0
  }

  at(s: number, out = new Vector3()): Vector3 {
    const n = this.pts.length
    if (!n) return out.set(0, 0, 0)
    if (n === 1 || s <= 0) return out.copy(this.pts[0])
    if (s >= this.length) return out.copy(this.pts[n - 1])
    let lo = 0, hi = n - 1                         // binary search for the segment containing s
    while (hi - lo > 1) { const mid = (lo + hi) >> 1; if (this.cum[mid] <= s) lo = mid; else hi = mid }
    const seg = this.cum[hi] - this.cum[lo]
    return out.lerpVectors(this.pts[lo], this.pts[hi], seg > 0 ? (s - this.cum[lo]) / seg : 0)
  }

  /** Horizontal unit direction of travel around s, or `fallback` if the path is degenerate there. */
  heading(s: number, back = 15, ahead = 40, fallback = new Vector3(0, 0, -1)): Vector3 {
    const a = this.at(s - back), b = this.at(s + ahead)
    const d = new Vector3(b.x - a.x, 0, b.z - a.z)
    return d.lengthSq() < 1e-6 ? fallback.clone() : d.normalize()
  }
}
